docs/compilers/CSharp/Nullability Public API Design Notes.md

```diff
@@ -0,0 +1,182 @@
+* `MaybeNull` - This is used when we know the nullability might be null.
+
+Nullability information will be retreived from the `SemanticModel`, much like type information. `GetSemanticModel` will return a `SemanticModel` that is aware of nullability information, and will force binding and calculation of nullability information when queried. We will offer a method off of `SemanticModel`, `GetSingleStatementSemanticModel()`, which will return a `SemanticModel` that does not calculate nullability information, for particularly performance-sensitive scenarios. This `SemanticModel` will share a cache with its parent, much like speculative `SemanticModel`s, and all nullabilities accessed on symbols retrieved from this `SemanticModel` will be `NotAnalyzed`.
+
+Generally speaking, when you want to determine the nullability of an `ITypeSymbol`, you need to check the containing context. The declaring symbol (e.g. `ILocalSymbol`, `IFieldSymbol`, etc.) will have the declared nullability of the type. `GetTypeInfo` will give you both the `ITypeSymbol` and the `Nullability` of an expression. Nested nullabilities for type parameters are contained on the containing `ITypeSymbol`. For example, if you wanted to know whether an array's contents are nullable, you will need to check the `IArrayTypeSymbol` for `ElementNullability`.
+
+### Declared Nullability
+
+Declared nullability is the nullability that was explicitly typed in the source code by the programmer. There are some cases where this is inferred, such as lambda arguments/return types, `var` variables, and type parameter substitution. For these cases, flow state is necessary to compute the declared nullability. For example:
+
+```C#
+string? s1 = string.Empty;
+var s2 = s1;
+s2 = null; // This will warn, because the type of s2 is string, not string?
+```
+
+The declared nullability of `s1` is `MaybeNull`, even though it is being initialized with a non null value. However, because of the flow state of `s1` is `NotNull`, we will infer the declared nullability of `s2` to be `NotNull` as well.
+
+Declared nullability can be retreived from the declaration symbol (e.g. `IFieldSymbol`, `IMethodSymbol`, etc.).
+
+### Flow State
+
+When the nullable feature is enabled, Roslyn will track the flow state of expressions throughout a method, regardless of what the variable was declared as. `SemanticModel` can be used to request flow state information for individual expressions or sub-expressions through the `GetTypeInfo` APIs. The `TypeInfo` struct will be augmented with `Nullability` and `ConvertedNullability` information, similar to how it currently has `Type` and `ConvertedType`. Nullability of symbols at a location can be retrieved through the `GetSymbolInfo` API: if the expression successfully resolved to a single symbol, the `Symbol` property will have nullability calculated. `CandidateSymbols` will not have calculated nullabilities, as we cannot run nullable analysis on code that has failed overload resolution.
```

Comment:
Does that mean CandidateSymbols will still have declared accessibilities or will be marked with `NotAnalyzed` to explicitly state it wasn't analyzed?

Reply:
CandidateSymbols will likely be `NotAnalyzed`. If overload resolution did not succeed we cannot run nullability analysis.

---
In reply to: [240431601](https://github.com/333fred/roslyn/pull/1#discussion_r240431601) [](ancestors = 240431601)